pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go

```diff
@@ -0,0 +1,210 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apibindingdeletion
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/reconciler/tenancy/clusterworkspacedeletion/deletion"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+const (
+	DeletionRecheckEstimateSeconds = 5
+
+	// ResourceDeletionFailedReason is the reason for condition BindingResourceDeleteSuccess that deletion of
+	// some CRs is failed
+	ResourceDeletionFailedReason = "ResourceDeletionFailed"
+
+	// ResourceRemainingReason is the reason for condition BindingResourceDeleteSuccess that some CR resource still
+	// exists when apibinding is deleting
+	ResourceRemainingReason = "SomeResourcesRemain"
+
+	// ResourceFinalizersRemainReason is the reason for condition BindingResourceDeleteSuccess that finalizers on some
+	// CRs still exist.
+	ResourceFinalizersRemainReason = "SomeFinalizersRemain"
+)
+
+type gvrDeletionMetadata struct {
+	// numRemaining is how many instances of the gvr remain
+	numRemaining int
+	// finalizersToNumRemaining maps finalizers to how many resources are stuck on them
+	finalizersToNumRemaining map[string]int
+}
+
+func (c *Controller) deleteAllCRs(ctx context.Context, apibinding *apisv1alpha1.APIBinding) error {
+
+	gvrToNumRemaining := map[schema.GroupVersionResource]int{}
+	finalizersToNumRemaining := map[string]int{}
+
+	deleteContentErrs := []error{}
+	for _, resource := range apibinding.Status.BoundResources {
+		for _, version := range resource.StorageVersions {
+			gvr := schema.GroupVersionResource{
+				Group:    resource.Group,
+				Resource: resource.Resource,
+				Version:  version,
+			}
+
+			deletionMetadata, err := c.deleteAllCR(ctx, logicalcluster.From(apibinding), gvr)
+			if err != nil {
+				deleteContentErrs = append(deleteContentErrs, err)
+			}
+
+			if deletionMetadata.numRemaining > 0 {
+				gvrToNumRemaining[gvr] = deletionMetadata.numRemaining
+				for finalizer, numRemaining := range deletionMetadata.finalizersToNumRemaining {
+					if numRemaining == 0 {
+						continue
+					}
+					finalizersToNumRemaining[finalizer] = finalizersToNumRemaining[finalizer] + numRemaining
+				}
+			}
+		}
+	}
+
+	if len(deleteContentErrs) > 0 {
+		conditions.MarkFalse(
```

Comment:
This side-effect mutation makes the logical flow hard to follow. Can we make this more explicit?